training/metrics.py

```diff
@@ -24,92 +21,244 @@ def count_trainable_params(model):
       Total number ot trainable parameters
   """
   weights = model.trainable_weights
-  total_trainable_params = int(sum(np.prod(p.shape.as_list()) for p in object_identity.ObjectIdentitySet(weights)))
+  total_trainable_params = int(sum(np.prod(tf.size(p)) for p in object_identity.ObjectIdentitySet(weights)))
   return total_trainable_params
 
+def _count_linear_layer(layer, num_of_blades):
+  """
+  Note: This function calculates the flops for any linear layer
 
-def _count_flops_conv2d(layer):
-  if layer.data_format == "channels_first":
-    input_channels = layer.input_shape[1]
-    output_channels, h, w, = layer.output_shape[1:]
-  elif layer.data_format == "channels_last":
-    input_channels = layer.input_shape[3]
-    h, w, output_channels = layer.output_shape[1:]
-  w_h, w_w = layer.kernel_size
-
-  n_mult = h * w * output_channels * input_channels * w_h * w_w
-  n_add = w_h * w_w * input_channels * h * w * output_channels
+  Args:
+    layer (tf.keras.layer.Conv2D or DepthwiseConv2D or Dense): any linear layer from the keras model
+    num_of_blades: This value represent the number of blades we use for different upstride types.
 
-  flops = n_mult + n_add
+  Returns:
+      int : total FLOPs for the given layer.
+  """
+  input_shape = layer.input_shape
+  output_shape = layer.output_shape
+  output_group = np.prod(output_shape[1:])
+  class_name = layer.__class__.__name__
+
+  # we get output image size (H*W) (this is channel first/last dependent)
+  if len(input_shape) == 4: # 2D Conv and DepthWise
+    # get the channel index
+    channel_index = 1 if layer.data_format == "channels_first" else -1
+    # get the input channels and output channels
+    input_channels = input_shape[channel_index]
+    output_channels = output_shape[channel_index]
+
+  # we get kernel size (this is layer dependant (Dense vs Conv))
+  # Depthwise layer output channels are same as input.
+  # It is a generalized case of group convolution where input channels are same as number of groups
```

Comment:
I don't understand this comment, or more precisely, I don't see why this comment is useful ?

Reply:
This was added to provide information regarding Conv2D and DepthwiseConv2D. I agree this is not relevant especially here.

Reply:
Do we never use the layer.groups ? I thought we would get the number of groups, with groups=input_channels for depthwise since it does not have a groups attribute, and then do the generic `operation flops += output_group * num_of_blades / groups` ?

Reply:
DepthwiseConv2D has groups but its always 1. so we have to rely on the output_channels or input_channels.